src/app/mod.rs

```diff
@@ -52,42 +52,50 @@ type Shared<T> = Arc<RwLock<T>>;
 #[derive(Clone)]
 pub(crate) struct BaseCoinApp<S> {
     store: MainStore<S>,
-    pub modules: Shared<Vec<Box<dyn Module<Store = ModuleStore<S>> + Send + Sync>>>,
-    account: Shared<BaseAccount>, // TODO(hu55a1n1): get from user and move to provable store
+    modules: Shared<ModuleList<S>>,
 }
 
 impl<S: Default + ProvableStore + 'static> BaseCoinApp<S> {
     /// Constructor.
     pub(crate) fn new(store: S) -> Result<Self, S::Error> {
-        let store = SharedStore::new(RevertibleStore::new(store));
-        // `SubStore` guarantees modules exclusive access to all paths in the store key-space.
-        let modules: Vec<Box<dyn Module<Store = ModuleStore<S>> + Send + Sync>> = vec![
-            Box::new(Bank::new(SubStore::new(
-                store.clone(),
-                prefix::Bank {}.identifier(),
-            )?)),
-            Box::new(Ibc::new(SubStore::new(
-                store.clone(),
-                prefix::Ibc {}.identifier(),
-            )?)),
-        ];
         Ok(Self {
-            store,
-            modules: Arc::new(RwLock::new(modules)),
-            account: Default::default(),
+            store: SharedStore::new(RevertibleStore::new(store)),
+            modules: Arc::new(RwLock::new(vec![])),
         })
     }
+
+    #[inline]
+    fn is_unique_id(&self, prefix: &Identifier) -> bool {
+        !self
+            .modules
+            .read()
+            .unwrap()
+            .iter()
+            .any(|(id, _)| id == prefix)
+    }
+
+    pub(crate) fn add_module(
+        self,
+        prefix: Identifier,
+        module: impl Module<ModuleStore<S>> + 'static,
+    ) -> Self {
+        assert!(self.is_unique_id(&prefix), "module prefix must be unique");
+        self.modules
+            .write()
+            .unwrap()
+            .push((prefix, Box::new(module)));
+        self
+    }
 }
 
 impl<S: Default + ProvableStore> BaseCoinApp<S> {
-    pub(crate) fn get_store(&self, prefix: Identifier) -> Option<ModuleStore<S>> {
-        let mut modules = self.modules.write().unwrap();
-        for m in modules.iter_mut() {
-            if m.store().prefix() == prefix {
-                return Some(m.store().clone());
-            }
-        }
-        None
+    pub(crate) fn module_store(&self, prefix: &Identifier) -> SharedStore<ModuleStore<S>> {
+        let modules = self.modules.read().unwrap();
+        modules
+            .iter()
+            .find(|(p, _)| p == prefix)
+            .map(|(_, m)| m.store().clone())
```

Comment:
Could we have a method on the `SharedStore` that delegates to `clone()` but is a more descriptive name, such as `get_rc_reference()` (or anything better you can think of)?

It wasn't clear to me why we would clone `m.store()`; the fact that it's an `Arc` (which has a "special" semantic to `clone()`) is hidden behind the `SharedStore` definition; and the fact that it's a `SharedStore` at all is not clear from just reading `m.store()`.

Reply:
Good point! Let's do `share()`. 👍

Reply:
https://github.com/informalsystems/basecoin-rs/pull/44/commits/c0bb6667f0881b29520b1044a0cbd0d5c85a22dc